File: app/src/main/java/home/oleg/placesnearme/presentation/feature/map/view/VenuesMapFragment.java
package home.oleg.placesnearme.presentation.feature.map.view;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.view.View;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;
import com.smedialink.common.Optional;
import com.smedialink.common.Pair;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;

import home.oleg.placenearme.models.UserLocation;
import home.oleg.placesnearme.PlacesNearMeApp;
import home.oleg.placesnearme.R;
import home.oleg.placesnearme.di.HasComponent;
import home.oleg.placesnearme.presentation.feature.map.di.PlacesMapFragmentComponent;
import home.oleg.placesnearme.presentation.feature.map.marker.MarkerMapper;
import home.oleg.placesnearme.presentation.feature.map.viewmodel.VenuesMapViewModelFacade;
import home.oleg.placesnearme.presentation.feature.venue.view.VenueDetailsView;
import home.oleg.placesnearme.presentation.viewdata.VenueViewData;
import permissions.dispatcher.NeedsPermission;
import permissions.dispatcher.RuntimePermissions;

import static home.oleg.placesnearme.presentation.feature.map.view.VenuesMapFragmentPermissionsDispatcher.initLocationSettingsWithPermissionCheck;
import static home.oleg.placesnearme.presentation.feature.map.view.VenuesMapFragmentPermissionsDispatcher.onShowCurrenLocationClickedWithPermissionCheck;

@RuntimePermissions
public class VenuesMapFragment extends BaseMapFragment implements
        VenuesMapView,
        GoogleMap.OnMarkerClickListener,
        HasComponent<PlacesMapFragmentComponent>,
        GoogleMap.OnMapClickListener {

    private static final int USER_LOCATION_ZOOM = 16;

    @Inject
    VenuesMapViewModelFacade venuesMapViewModelFacade;

    @Inject
    MarkerMapper markerMapper;

    private PlacesMapFragmentComponent component;

    private GoogleMap googleMap;

    private VenueDetailsView venueDetailsView;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        injectDependencies();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        venueDetailsView = view.findViewById(R.id.venueView);

        FloatingActionButton fabCurrentLocation = view.findViewById(R.id.fabCurrentLocation);
        FloatingActionButton fabZoomIn = view.findViewById(R.id.fabZoomIn);
        FloatingActionButton fabZoomOut = view.findViewById(R.id.fabZoomOut);

        fabCurrentLocation.setOnClickListener(v ->
                onShowCurrenLocationClickedWithPermissionCheck(this));

        fabZoomIn.setOnClickListener(v -> {
            Optional.of(googleMap)
                    .ifPresent(m -> m.animateCamera(CameraUpdateFactory.zoomIn()));
        });

        fabZoomOut.setOnClickListener(v -> {
            Optional.of(googleMap)
                    .ifPresent(m -> m.animateCamera(CameraUpdateFactory.zoomOut()));
        });

        venuesMapViewModelFacade.attachView(this);
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        this.googleMap = googleMap;
        initLocationSettingsWithPermissionCheck(this, googleMap);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        VenuesMapFragmentPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults);
    }

    @Override
    public void showVenues(@NonNull List<VenueViewData> items) {
        Optional.of(googleMap).ifPresent(map -> {
            map.clear();

            List<Pair<MarkerOptions, VenueViewData>> pairs = markerMapper.mapFrom(items);
            Map<String, VenueViewData> markerVenueViewDataMap = new HashMap<>();

            for (Pair<MarkerOptions, VenueViewData> pair : pairs) {
                String id = map.addMarker(pair.getFirst()).getId();
                markerVenueViewDataMap.put(id, pair.getSecond());
            }

            venuesMapViewModelFacade.setVenues(markerVenueViewDataMap);
        });
    }

    @Override
    public void showUserLocation(UserLocation userLocation) {
        LatLng latLng = new LatLng(userLocation.getLat(), userLocation.getLng());

        Optional.of(googleMap).ifPresent(googleMap -> {
            googleMap.animateCamera(CameraUpdateFactory.newLatLngZoom(latLng, USER_LOCATION_ZOOM));
        });
    }

    @Override
    public void showError() {

    }

    @Override
    public void showLoading() {

    }

    @Override
    public void hideLoading() {

    }

    @SuppressLint("MissingPermission")
    @NeedsPermission(Manifest.permission.ACCESS_COARSE_LOCATION)
    void onShowCurrenLocationClicked() {
        venuesMapViewModelFacade.refreshUserLocation();
    }

    @SuppressLint("MissingPermission")
    @NeedsPermission(Manifest.permission.ACCESS_COARSE_LOCATION)
    void initLocationSettings(GoogleMap googleMap) {
        googleMap.getUiSettings().setMyLocationButtonEnabled(false);
        googleMap.setMyLocationEnabled(true);
        googleMap.setOnMarkerClickListener(this);
        googleMap.setOnMapClickListener(this);

        venuesMapViewModelFacade.refreshUserLocation();
        venuesMapViewModelFacade.refreshRecomendedVenues();
    }

    private void injectDependencies() {
       /* component = DaggerApplicationComponent.builder()
                .bind((PlacesNearMeApp) getActivity().getApplication())
                .build()
                .placeMapFragmentComponentBuilder()
                .bind(this)
                .build();
        component.inject(this);*/
    }

    @Override
    public boolean onMarkerClick(Marker marker) {
        venuesMapViewModelFacade.selectVenue(marker.getId());
        return true;
    }

    @Override
    public PlacesMapFragmentComponent get() {
        return component;
    }

    @Override
    public void onMapClick(LatLng latLng) {
        venuesMapViewModelFacade.geocode(latLng.latitude, latLng.longitude);
    }

    @Override
    public void showVenue(VenueViewData venueViewData) {
        venueDetailsView.showVenue(venueViewData);
    }
}

File: feature-map/src/main/java/home/oleg/placesnearme/feature_map/presentation/VenuesMapFragment.kt
package home.oleg.placesnearme.feature_map.presentation

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity.RESULT_OK
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.core.view.isVisible
import com.google.android.gms.common.api.ResolvableApiException
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.model.LatLng
import com.olegsheliakin.lifecycle.autodisposable.AutoDisposable
import com.olegsheliakin.lifecycle.autodisposable.disposeOnDestroy
import com.olegsheliakin.statebinder.StateBinder
import com.smedialink.common.base.BackHandler
import com.smedialink.common.base.BaseFragment
import com.smedialink.common.ext.*
import home.oleg.placesnearme.corepresentation.api.ShowHideBottomBar
import home.oleg.placesnearme.corepresentation.delegate.ToastDelegate
import home.oleg.placesnearme.corepresentation.viewdata.PreviewPlace
import home.oleg.placesnearme.coredata.location.ReactiveLocationSettings
import home.oleg.placesnearme.coredomain.models.Section
import home.oleg.placesnearme.feature_map.R
import home.oleg.placesnearme.feature_map.di.PlacesMapFragmentComponent
import home.oleg.placesnearme.feature_map.mapper.MarkerMapper
import home.oleg.placesnearme.feature_map.presentation.adapter.SectionsAdapter
import home.oleg.placesnearme.feature_map.presentation.ui.initLocationSettingsWithPermissionCheck
import home.oleg.placesnearme.feature_map.presentation.ui.onRequestPermissionsResult
import home.oleg.placesnearme.feature_map.presentation.ui.onShowCurrentLocationClickedWithPermissionCheck
import home.oleg.placesnearme.feature_venue_detail.presentation.VenueViewFacade
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.rxkotlin.subscribeBy
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.fragment_map.*
import kotlinx.android.synthetic.main.layout_search.*
import permissions.dispatcher.NeedsPermission
import permissions.dispatcher.RuntimePermissions
import javax.inject.Inject

private const val REQUEST_CODE_CHECK_LOCATION_SETTINGS = 10001

@RuntimePermissions
class VenuesMapFragment : BaseFragment(), SectionsAdapter.SectionSelectListener, BackHandler {

    override val layoutRes: Int = R.layout.fragment_map

    @Inject
    lateinit var toastDelegate: ToastDelegate

    @Inject
    lateinit var viewModel: MapViewModel

    @Inject
    lateinit var venueViewFacade: VenueViewFacade

    @Inject
    lateinit var markerMapper: MarkerMapper

    @Inject
    lateinit var sectionsAdapter: SectionsAdapter

    @Inject
    lateinit var reactiveLocationSettings: ReactiveLocationSettings

    private lateinit var showHideBottomBarListener: ShowHideBottomBar

    private val lifeCycleAwareMap = LifeCycleAwareMap.create(this, OnMapReadyCallback {
        this.googleMap = it

        googleMap?.setOnMarkerClickListener {
            viewModel.venuesHolder[it.id]?.let(venueViewFacade::setVenue)
            return@setOnMarkerClickListener true
        }

        viewModel.state.observe(this, stateBinder::newState)

        initLocationSettingsWithPermissionCheck(it)
    })

    private var autoDisposable = AutoDisposable.create(viewLifecycleOwner)

    private var stateBinder = StateBinder.create<MapViewState>()

    private var googleMap: GoogleMap? = null

    override fun onAttach(context: Context?) {
        super.onAttach(context)

        if (context is ShowHideBottomBar) {
            this.showHideBottomBarListener = context
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        lifeCycleAwareMap.onViewCreated(view, savedInstanceState)
        toastDelegate.attach(view.context)
        venueViewFacade.onCreateView(view, viewLifecycleOwner)
        venueViewFacade.setShowHideBottomBarListener(showHideBottomBarListener)

        stateBinder.applyCurrentState()

        sectionsAdapter.setHasStableIds(true)
        rvSections.horizontal(withAdapter = sectionsAdapter)

        bindClicks()
        bindState()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        val shouldShowSearch = searchAppBar.visibility == View.VISIBLE
        outState.putBoolean(KEY_SEARCH_VISIBILITY, shouldShowSearch)
        outState.putInt(KEY_SEARCH_SELECTED_ITEM_POSITION, sectionsAdapter.selectedItemPosition)
        venueViewFacade.onSaveState(outState)
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        super.onViewStateRestored(savedInstanceState)
        if (savedInstanceState != null) {
            val position = savedInstanceState.getInt(KEY_SEARCH_SELECTED_ITEM_POSITION)
            val shouldShow = savedInstanceState.getBoolean(KEY_SEARCH_VISIBILITY)
            if(shouldShow) {
                viewModel.openSearch()
            } else {
                viewModel.closeSearch()
            }
            sectionsAdapter.setSelected(position)
            venueViewFacade.onRestoreState(savedInstanceState)
        } else {
            showSearch(false)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        onRequestPermissionsResult(requestCode, grantResults)
    }

    private fun bindClicks() {
        loadingView.setOnRetryCLickListener {
            sectionsAdapter.selectedItem?.let(viewModel::getRecommendedVenues)
        }

        fabSearch.setOnClickListener {
            viewModel.openSearch()
        }

        btnClose.setOnClickListener {
            viewModel.closeSearch()
        }

        fabZoomIn.setOnClickListener {
            googleMap?.animateCamera(CameraUpdateFactory.zoomIn())
        }

        fabZoomOut.setOnClickListener {
            googleMap?.animateCamera(CameraUpdateFactory.zoomOut())
        }

        fabCurrentLocation.setOnClickListener {
            onShowCurrentLocationClickedWithPermissionCheck()
        }

    }

    private fun bindState() {
        stateBinder.apply {
            bind(MapViewState::isVenuesLoading) {
                loadingView.showLoading(it)
            }

            bind(MapViewState::isSearchShown, this@VenuesMapFragment::showSearch)

            bindNullable(MapViewState::error, toastDelegate::onChanged)

            bind(MapViewState::places, this@VenuesMapFragment::renderPlaces)

            bindNullable(MapViewState::userLatLng) {
                it?.let(this@VenuesMapFragment::renderLocation)
            }
        }
    }

    private fun renderPlaces(items: Collection<PreviewPlace>) {
        googleMap?.apply {
            clear()

            val pairs = markerMapper.mapFrom(items)
            val markerVenueViewDataMap = HashMap<String, PreviewPlace>()

            for (pair in pairs) {
                val id = addMarker(pair.first).id
                markerVenueViewDataMap[id] = pair.second
            }

            viewModel.venuesHolder.clear()
            viewModel.venuesHolder.putAll(markerVenueViewDataMap)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_CODE_CHECK_LOCATION_SETTINGS && resultCode == RESULT_OK) {
            viewModel.requestUserLocation()
        }
    }

    @SuppressLint("MissingPermission")
    @NeedsPermission(Manifest.permission.ACCESS_FINE_LOCATION)
    fun onShowCurrentLocationClicked() {
        viewModel.requestUserLocation()
    }

    @SuppressLint("MissingPermission")
    @NeedsPermission(Manifest.permission.ACCESS_FINE_LOCATION)
    fun initLocationSettings(googleMap: GoogleMap) {
        reactiveLocationSettings.checkSettings()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(
                        onComplete = {
                            viewModel.requestUserLocation()
                            googleMap.apply {
                                uiSettings.isMyLocationButtonEnabled = false
                                isMyLocationEnabled = true
                            }
                        },
                        onError = {
                            if (it is ResolvableApiException) {
                                it.startResolutionForResult(activity, REQUEST_CODE_CHECK_LOCATION_SETTINGS)
                            } else {
                                toastDelegate.showError(getString(R.string.unexpected_error))
                            }
                        }
                ).disposeOnDestroy(autoDisposable)
    }

    override fun sectionSelected(section: Section) {
        viewModel.getRecommendedVenues(section)
    }

    override fun inject() {
        PlacesMapFragmentComponent.Injector.inject(this)
    }

    override fun onLowMemory() {
        super.onLowMemory()
        lifeCycleAwareMap.onLowMemory()
    }

    override fun onBackPressed(): Boolean {
        if (venueViewFacade.dismiss()) {
            return true
        }

        if (searchAppBar.isVisible) {
            viewModel.closeSearch()
            return true
        }

        return false
    }

    private fun showSearch(shouldShow: Boolean) {
        searchAppBar.gone(!shouldShow)
        fabSearch.showExt(shouldShow)
    }

    private fun renderLocation(userLocation: home.oleg.placesnearme.coredomain.models.LatLng) {
        val latLng = LatLng(userLocation.lat, userLocation.lng)
        googleMap?.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, USER_LOCATION_ZOOM.toFloat()))
    }

    companion object {
        private const val KEY_SEARCH_VISIBILITY = "key_search_visibility"
        private const val KEY_SEARCH_SELECTED_ITEM_POSITION = "key_search_selected_item_position"
        private const val USER_LOCATION_ZOOM = 14
    }
}
